src/system/user.cr

```diff
@@ -0,0 +1,201 @@
+require "c/pwd"
+require "c/sys/limits"
+
+struct System::User
+  module Limits
+    {% if flag?(:darwin) || flag?(:openbsd) || flag?(:freebsd) %}
+      UID_MAX = LibC::UID_MAX
+    {% elsif flag?(:linux) %}
+      UID_MAX = 0xffffffff_u32
+    {% else %}
+      UID_MAX = 0xffffffff_u32
+    {% end %}
+  end
+
+  # Converts user ID into a username.
+  #
+  # Returns: The username for the given user ID.
+  #
+  # ```
+  # System::User.name(0) # => root
+  # ```
+  def self.name(uid : Int) : String
+    user_struct = LibC.getpwuid(int_to_uid(uid))
+    return String.new(user_struct.value.pw_name) if !user_struct.null?
+    raise NotFound.new("User with uid '#{uid}', was not found.")
+  end
+
+  # Converts username into user ID.
+  #
+  # Returns: The user ID for the given username.
+  # Raises: User::NotFound error if no user exists with the given username.
+  #
+  # ```
+  # System::User.uid("root") # => 0
+  # ```
+  def self.uid(user : String) : Int
+    user_struct = LibC.getpwnam(user.check_no_null_byte)
+    return user_struct.value.pw_uid if !user_struct.null?
+    raise NotFound.new("User with name '#{user}', was not found.")
+  end
+
+  # Checks to see if user with a given user ID exists.
+  #
+  # Returns: `true` if the user exists.
+  #
+  # ```
+  # System::User.exists?("root")             # => true
+  # System::User.exists?("nonexistant_user") # => false
+  # ```
+  def self.exists?(uid : Int) : Bool
+    !LibC.getpwuid(int_to_uid(uid)).null?
+  end
+
+  # Checks to see if user with a given username exists.
+  #
+  # Returns: `true` if the user exists.
+  #
+  # ```
+  # System::User.exists?(0)     # => true
+  # System::User.exists?(32766) # => false
+  # ```
+  def self.exists?(user : String) : Bool
+    !LibC.getpwnam(user.check_no_null_byte).null?
+  end
+
+  # Returns the user specified by the user ID.
+  # ```
+  # System::User.get(0)
+  # ```
+  def self.get(uid : Int) : User
+    new(uid)
+  end
+
+  # Returns the user specified by the username.
+  # ```
+  # System::User.get("root")
+  # ```
+  def self.get(username : String) : User
+    new(username)
+  end
+
+  # Initializes a user by the given user ID.
+  def initialize(uid : Int)
+    user_struct = LibC.getpwuid(User.int_to_uid(uid))
+    raise NotFound.new("User with uid '#{uid}', was not found.") if user_struct.null?
```

Comment:
`if user_struct.null?` => `unless user_struct`

Reply:
Fixed to my standards. I will not use `unless`.

Reply:
ditto